### PR TITLE
chore: collect samples per second for pytorch and keras [DET-5515]

### PR DESCRIPTION
**Pull request by stoksc**
## Description
This change adds "misc" metrics and collects samples_per_second as the first one, so we can display it in the webui again, but not rely on getting it from a step's metrics (since those will go away shortly). It also heavily refactors the profiler code to have a single "batcher" type and a single "batcher thread" type, instead of one specifically for timings/system/misc metrics.

## Test Plan
- [x] run e2es written for profiling,
- [x] run the code and made sure samples per second are collected.

## Commentary (optional)
This code is not yet "always on". I'm going to leave that for a follow up PR / discuss it at ml-ag.

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: harness/determined/profiler.py [hunk 1]
```diff
@@ -581,50 +630,55 @@ def run(self) -> None:
...
-                self.send_queue.put(self.current_batch.convert_to_post_format())
-                self.current_batch.clear()
+                self.send_queue.put(self.current_batch.consume())
                 batch_start_time = time.time()
 
 
-class TimingsBatcherThread(threading.Thread):
+class MetricsBatcherThread(threading.Thread):
     """
     This is a thread that exists solely so that we can batch Timings and ship them to the
```
Comment (armandmcqueen) [inline harness/determined/profiler.py:683]: Out-of-date comment

File: harness/determined/profiler.py [hunk 2]
```diff
@@ -704,11 +758,87 @@ def run(self) -> None:
...
                 batch_start_time = time.time()
 
 
+class MetricBatch:
+    def __init__(self, trial_id: str, agent_id: str) -> None:
+        self.trial_id = trial_id
+        self.agent_id = agent_id
+        self.batch = {}  # type: Dict[Tuple[MetricType, str, str], List[Measurement]]
+
+    def append(
```
Comment (armandmcqueen) [inline harness/determined/profiler.py:771]: nit: I personally wouldn't call this `append` because `append` is a threadsafe operation in Python while this operation is not. Not very important, so feel free to leave it as `append` if you think that's the best name.

File: harness/determined/profiler.py [hunk 3]
```diff
@@ -704,11 +758,87 @@ def run(self) -> None:
...
-                self.current_batch.clear()
+                self.send_queue.put(self.metrics_batch.consume())
                 batch_start_time = time.time()
 
 
+class MetricBatch:
+    def __init__(self, trial_id: str, agent_id: str) -> None:
+        self.trial_id = trial_id
+        self.agent_id = agent_id
+        self.batch = {}  # type: Dict[Tuple[MetricType, str, str], List[Measurement]]
```
Comment (armandmcqueen) [inline harness/determined/profiler.py:769]: This is a much better design!

**Review by armandmcqueen — CHANGES_REQUESTED** [review]
The code looks good, but it's missing the backend changes.

**Comment by stoksc** [comment]
> The code looks good, but it's missing the backend changes.

Which backend changes exactly?

**Comment by armandmcqueen** [comment]
> > The code looks good, but it's missing the backend changes.
> 
> Which backend changes exactly?

The API's ability to accept the new MISC type of metric. Don't you need to add the MISC type [here](https://github.com/determined-ai/determined/blob/master/proto/src/determined/trial/v1/trial.proto#L104)?

**Review by armandmcqueen — APPROVED** [review]
